### ISSUE TITLE
Calling Java methods from javascript

### ISSUE THREAD

**Issue by jamieJames45** [CodeBrig/Journey#25]
Is this currently planned functionality?
is CefMessageRouter the ideal?

**Comment by BFergerson** [comment]
I believe a fork of Journey has implemented this functionality (https://github.com/dhruvit-r/Journey). Maybe give that a try first and if it works we can merge those changes back here.

**Comment by jamieJames45** [comment]
Thanks!  I will take a look and get back to you.

**Comment by jamieJames45** [comment]
reporting back, 

I was unable to get that one to Build. 
'class file for org.joor.Reflect not found' error when trying to build. 
Issue looks related to  this dependancy, compileOnly 'com.github.bfergerson:joor:bef9254b2a'

**Comment by BFergerson** [comment]
@jamieJames45, I wasn't expecting you to build it. The author has already provided a release (https://github.com/dhruvit-r/Journey/releases). Using that release would be the simplest way of verifying if it does what you're looking for.

**Comment by jamieJames45** [comment]
My apologies for the misunderstanding. Thanks for the response.

**Comment by jamieJames45** [comment]
Granted,  I know you didn't write this part,  but since we are discussing, and fo your info, attempting to run this as an example:
https://github.com/dhruvit-r/Journey/blob/master/src/test/java/com/codebrig/journey/JourneyBrowser.java

results in : 
`Exception in thread "main" java.lang.IllegalArgumentException: not a proxy instance
	at java.lang.reflect.Proxy.getInvocationHandler(Proxy.java:816)
	at com.codebrig.journey.proxy.browser.CefMessageRouterProxy.lambda$static$0(CefMessageRouterProxy.java:26)
	at org.joor.Reflect$1.invoke(Reflect.java:754)
	at com.sun.proxy.$Proxy3.addHandler(Unknown Source)`

**Comment by BFergerson** [comment]
@jamieJames45, it seems you're running the class directly without the required dependencies. I think it might be easier to just use include that fork's dependencies entirely and just run it like so: https://github.com/BFergerson/journey-dhruvit-r-fork-test

Running that I can't find any issues with the fork. I wasn't able to test if the JavaScript integration worked though because I'm not aware of how it should work to begin with. I can only say that the code appears to be correct and it appears to be functional.

It would be up to you now to write the appropriate JavaScript code to communicate with CEF/JCEF (https://bitbucket.org/chromiumembedded/cef/wiki/JavaScriptIntegration). If you're able to get that working I'll gladly merge in the fork and improve the documentation so others can easily do this as well.

**Comment by jamieJames45** [comment]
Apologies for not getting back to you sooner. I got pulled on to another project. 
I can verify that it works as expected. 
I can provide some samples if you like.
Thanks for your help.

**Comment by BFergerson** [comment]
@jamieJames45, don't worry about it. I would love a small reproducible example of the JavaScript integration if you got one.

**Comment by jamieJames45** [comment]
Based on example from the jcef wiki

MessageHandler class

```
class MessageHandler extends CefNativeDefault implements CefMessageRouterHandlerProxy {

    public void onQueryCanceled(CefBrowserProxy browser, CefFrameProxy frame, long query_id) {
        System.out.println("Query cancelled.");
    }

    public boolean onQuery(CefBrowserProxy browser, CefFrameProxy frame, long query_id, String request,
        boolean persistent, CefQueryCallbackProxy callback) {
        System.out.println("Query succeeded: " + request);
//TODO switch
        if (request.startsWith("javaVersion")) {
            // return something back to JS
            callback.success(System.getProperty("java.version"));
            return true;
        }
        if (request.startsWith("showDevTools")) {

            CefBrowserProxy devTools = browser.getDevTools();
            Component uiComponent = devTools.getUIComponent();
            JFrame devFrame = new JFrame("DevToolsWindow");
            devFrame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            devFrame.setSize(400, 400);

            //Adding Components to the frame.
            devFrame.getContentPane().add(BorderLayout.CENTER, uiComponent);
            devFrame.setVisible(true);

            return true;
        } else {
            return false;
        }
    }
}
```
On the JS side:
```
function execute(cmd) {
  //generic execute handler
            try {


                window.cefQuery({
                    request: cmd,
                    onSuccess: function (response) {
                      
                        return response;
                    },
                    onFailure: function (error_code, error_message) {
                        document.getElementById('result').value = 'Error: ' + error_message;
                    }
                });
            } catch (ex) {
                console.dir(ex);
            }
        }
// callable JS methods
        function showDevTools() {
            execute("showDevTools");
        }
        function getJavaVersion() {

          alert(execute("javaVersion"));
        }

```

**Comment by tokazio** [comment]
Hello, what the state of this issue? Planned to merge ?

**Comment by BFergerson** [comment]
@tokazio, thanks for reminding me about this. I've merged in changes for the JavaScript integration.
Take a look at:
 - https://github.com/CodeBrig/Journey/blob/master/src/test/java/com/codebrig/journey/JourneyBrowser.java
 - https://github.com/CodeBrig/Journey/blob/master/src/test/resources/JSTest.html

Running JourneyBrowser should bring up:
![Screenshot from 2019-12-01 02-00-50](https://user-images.githubusercontent.com/3278877/69911831-7a13b200-13de-11ea-996a-dc4487a9bf16.png)

p.s. I see your issue (#28) about running Journey. I'll look at that today.

---

@jamieJames45, were you actually able to get `alert(execute("javaVersion"));` to work? I kept running into `No javascript dialog runner available for this platform jcef`. Not a big deal if not. Just curious. Thanks for the example code btw.
